src/alarms/alarm-handling.md

```diff
@@ -0,0 +1,87 @@
+
+# Information Set per Alarm
+The specified alarm attributes below are inspired by cumulocity data model for alarms, but shall be re-usable for other clouds where possible.
+
+| Name              | Description   |
+| ----------------- |---------------|
+| ref-string        | Device-unique reference string (kind of alarm ID or alarm name), used to reference the once occured alarm again, e.g. in case of updates (e.g. "temperature_sensor_loss") |
+| message-string    | Human readable short information about alarm reason (e.g. "Temperature sensor does not respond") |
+| severity-string   | Could be "CRITICAL", "MAJOR", "MINOR" or "WARNING" |
+| status-string     | Could be "ACTIVE" or "CLEARED"<br/>
+| time-string       | Timestamp indicating when the alarm (or the alarm-update) had occured (in [ISO 8601 format](http://www.w3.org/TR/NOTE-datetime#)) |
+
+# Data flow
+
+Figure below illustrates the data flow from Customer Application broker/thin-edge up to cloud. It shows especially the use of MQTT QOS=1, MQTT Retain and Mosquitto Persistence to achieve reliability alarm state-change transfer.
+
+![Sequence Diagram Update SW-list](images/alarm_dataflow.svg)
+
+**TODO:** Interface from Mapper to Cloud to be defined. Two options are possible:
+1) **JSONviaMQTT:** To have reliability use QOS=1. Open issue: Not yet completely confirmed that C8Y sends PUBACK when msg was processed, instead just on arrival.
+2) **HTTP REST:** HTTP response indicates if message was processed sucessfully. Anyway HTTP REST would increase complexity of implementation and increase data traffic.
+
+
+# Public MQTT-based alarm interface
+
+Similar to thin-edge'S measurement interface, alarm interface is based on MQTT topics.
+
+## Topic structure and payload
+
+Proposal to have "severity" and "ref-string" as topics:
+
+**thin-edge JSON (alarm) format:**
+```
+Topic:   tedge/alarms/<device-id>/<severity-string>/<ref-string>
+Payload: {
+  "message":  <message-string>,
+  "status":   <status-string>,  
+  "time":     <time-string>
+}
+```
+
+NOTE: The "device-id" indicates whether it is an alarm of the thin-edge device it-self (root-device) or a child-device. Use "root" for the thin-edge device or the child-device id for the according child-device.
+
+Benefit to have "severity" and "ref-string" as topics:
+Device-Site reactions to alarm could be easier relized.<br/>
+Examples:
+```
+# Listen to all major alarms on all devices
+tedge mqtt sub "tedge/alarms/+/major/+"
+
+# Listen to specific alarm on all devices
+# (e.g. a bunch of temperature sensors that have each an upper limit setpoint with alarm)
+tedge mqtt sub "tedge/alarms/+/+/upper-limit-exceeded"
+
+# And for sure all combinations.
+```
+
+
+Additionally an also supported "raw"-payload format will allow easy adoption for existing "MQTT publishers":
+
+**Raw payload format:**
+```
+Topic:   tedge/alarms/<device-id>/<severity-string>/<ref-string>

```

Comment:
That's a good point.

Reply:
That's true. Here we have pros and cons for moving the severity into the topic structure. However putting severity into the topic make it possible in the first place to define the raw payload. Otherwise I see no way to do that.

By the way: Having a fixed severity for each alarm is the way I know from PLCs. E.g. in CoDeSys you can define an Alarm with just a fixed severity. I.e. in case of your temperature-monitoring example, CoDeSys would raise another alarm that has another severity.